internal/providers/zvm/zvm.go

```diff
@@ -0,0 +1,108 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The zVM provider fetches a local configuration from the virtual unit
+// record devices.
+
+package zvm
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"strings"
+
+	"github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/coreos/ignition/v2/internal/distro"
+	"github.com/coreos/ignition/v2/internal/log"
+	"github.com/coreos/ignition/v2/internal/providers/util"
+	"github.com/coreos/ignition/v2/internal/resource"
+	"github.com/coreos/vcontext/report"
+)
+
+const readerDevice string = "000c"
+
+func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	// Fetch config files directly from reader device.
+	_, err := f.Logger.LogCmd(exec.Command(distro.ModprobeCmd(), "vmur"), "Loading zVM control program module")
+	if err != nil {
+		f.Logger.Err("Couldn't install vmur module: %v", err)
+		errors := fmt.Errorf("Couldn't install vmur module: %v", err)
+		return types.Config{}, report.Report{}, errors
+	}
+	// Online the reader device.
+	logger := f.Logger
+	err = onlineDevice(logger)
+	if err != nil {
+		return types.Config{}, report.Report{}, err
+	}
+	// Read files from the z/VM reader queue.
+	readerInfo, err := exec.Command(distro.VmurCmd(), "li").CombinedOutput()
+	if err != nil {
+		f.Logger.Err("Can not get reader device: %v", err)
+		errors := fmt.Errorf("Can not get reader device: %v", err)
+		return types.Config{}, report.Report{}, errors
+	}
+	for _, records := range strings.Split(string(readerInfo), "\n") {
+		record := strings.Fields(records)
+		if len(record) <= 1 {
+			break
+		}
+		if len(record) < 11 {
+			continue
+		}
+		spoolid := record[1]
+		ftype := record[10]
+		file := record[9] + "." + record[10]
+		// Receive the spool file.
+		if ftype == "ign" {
+			_, err := f.Logger.LogCmd(exec.Command(distro.VmurCmd(), "re", "-f", spoolid, file), "Receive the spool file")
+			if err != nil {
+				return types.Config{}, report.Report{}, err
+			}
+			f.Logger.Info("using config file at %q", file)
+			rawConfig, err := ioutil.ReadFile(file)
+			if err != nil {
+				f.Logger.Err("Couldn't read config from configFile %q: %v", file, err)
+				break
+			}
+			jsonConfig := bytes.Trim(rawConfig, string(byte(0)))
+			return util.ParseConfig(f.Logger, jsonConfig)
+		}
+	}
+	return types.Config{}, report.Report{}, errors.ErrEmpty
+}
+
+func onlineDevice(logger *log.Logger) error {
+	_, err := logger.LogCmd(exec.Command(distro.ChccwdevCmd(), "-e", readerDevice), "Brings a Linux device online")
+	if err != nil {
```

Comment:
Is there a special convention e.g. exit code we could here here to detect the case where it's worth retrying?

Alternatively, is `cio_ignore -r` idempotent? Could we just always do it upfront before running `chccwdev`?

Reply:
> Is there a special convention e.g. exit code we could here here to detect the case where it's worth retrying?
> Alternatively, is `cio_ignore -r` idempotent? Could we just always do it upfront before running `chccwdev`?

This command set attributes for CCW devices and to set CCW devices online. If it failed , we use "cio_ignore -r " to remove device specifications from the exclusion list, the corresponding devices are sensed and analyzed if they exist, and the devices become available to Linux. So The "cio_ignore -r " means retrying, If the retry failed, it will failed to online the device.

Reply:
> If it failed , we use "cio_ignore -r " to remove device specifications from the exclusion list, the corresponding devices are sensed and analyzed if they exist, and the devices become available to Linux.

Hmm, so it sounds like we could just always call `cio_ignore -r`, no? If it's being excluded, then it's removed from the list, if it's not... I assume it's just a no-op?

Anyway, don't know much about these s390x commands, so feel free to ignore my suggestions if it doesn't make sense. :)